spacy_llm/tasks/entity_linking.py

```diff
@@ -0,0 +1,310 @@
+
+    def generate_prompts(self, docs: Iterable[Doc]) -> Iterable[str]:
+        environment = jinja2.Environment()
+        _template = environment.from_string(self._template)
+        for doc in docs:
+            cands_ents_info, _ = self._fetch_entity_info(doc)
+            yield _template.render(
+                text=EntityLinkingTask._highlight_ents_in_text(doc),
+                mentions_str=", ".join([f"*{mention}*" for mention in doc.ents]),
+                mentions=[ent.text for ent in doc.ents],
+                entity_descriptions=[
+                    cands_ent_info[1] for cands_ent_info in cands_ents_info
+                ],
+                entity_ids=[cands_ent_info[0] for cands_ent_info in cands_ents_info],
+                examples=self._prompt_examples,
+            )
+
+    def parse_responses(
+        self, docs: Iterable[Doc], responses: Iterable[str]
+    ) -> Iterable[Doc]:
+        for doc, prompt_response in zip(docs, responses):
+            solutions = [
+                sol.replace("--- ", "").replace("<", "").replace(">", "")
+                for sol in re.findall(r"--- <.*>", prompt_response)
+            ]
+            # Skip document if the numbers of entities and solutions don't line up.
+            if len(solutions) != len(doc.ents):
+                yield doc
+
+            # Set ents anew by copying them and specifying the KB ID.
+            doc.ents = [
+                Span(
+                    doc=doc,
+                    start=ent.start,
+                    end=ent.end,
+                    label=ent.label,
+                    vector=ent.vector,
+                    vector_norm=ent.vector_norm,
+                    kb_id=solution,
+                )
+                for ent, solution in zip(doc.ents, solutions)
+            ]
+
+            yield doc
+
+    def scorer(
+        self,
+        examples: Iterable[Example],
+    ) -> Dict[str, Any]:
+        """Scores lemmatization accuracy on provided examples.
+        examples (Iterable[Example]): Examples to determine score against.
+        """
+        return Scorer.score_links(examples, negative_labels=[EntityLinker.NIL])
+
+    @property
+    def _cfg_keys(self) -> List[str]:
+        return ["_template"]
+
+    @property
+    def _Example(self) -> Type[EntityLinkingExample]:
+        return EntityLinkingExample
+
+    def _create_prompt_example(
+        self, example: Example
+    ) -> Optional[EntityLinkingExample]:
+        """Create a entity linking prompt example from a spaCy example."""
+        # Ensure that all entities have their knowledge base IDs set.
+        n_ents = len(example.reference.ents)
+        n_set_kb_ids = sum([ent.kb_id != 0 for ent in example.reference.ents])
+        if n_ents and n_ents != n_set_kb_ids:
+            warnings.warn(
+                f"Not all entities in this document have their knowledge base IDs set ({n_set_kb_ids} out of "
+                f"{n_ents}). Ignoring example:\n{example.reference}"
```

Comment:
Why are we doing this check? I can imagine some gold data that has entities but not all entities having disambiguated links?

Reply:
Can't we recover from this - only send the disambiguated cases to the LLM as example, and remove the entities without a link from the prompt alltogether?